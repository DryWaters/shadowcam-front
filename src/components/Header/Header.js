import React, { Component } from "react";
import { connect } from "react-redux";
import {
  Container,
  Collapse,
  Navbar,
  NavbarToggler,
  Button,
  Nav,
  NavItem,
  NavLink,
  UncontrolledDropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem
} from "reactstrap";
import { NavLink as NvLink, Link } from "react-router-dom";
import { logout } from "../../store/actions/user";

import styles from "./Header.module.css";
import logo from "../../assets/images/boxer-header.png";

export class Header extends Component {
  state = {
    isOpen: false
  };

  toggle = () => {
    this.setState({
      isOpen: !this.state.isOpen
    });
  };

  render() {
    const loginDropdown = (
      <UncontrolledDropdown nav inNavbar>
        <DropdownToggle caret>Account</DropdownToggle>
        <DropdownMenu right>
          <DropdownItem className={styles.dropdownItem}>
            <Link to="/account/profile">Account Settings</Link>
          </DropdownItem>
          <DropdownItem
            className={styles.dropdownItem}
            onClick={this.props.logout}
          >
            Logout
          </DropdownItem>
        </DropdownMenu>
      </UncontrolledDropdown>
    );

    const unauthLogin = (
      <Link to="/account/login">
        <Button>Login</Button>
      </Link>
    );

    return (
      <Navbar className={styles.headerContainer} color="dark" dark expand="md">
        <Container>
          <Link className={styles.logoLink} to="/">
            <img src={logo} alt="Logo" />
            <span className={styles.logo}>ShadowCam</span>
          </Link>
          <NavbarToggler onClick={this.toggle} />
          <Collapse isOpen={this.state.isOpen} navbar>
            <Nav className={styles.menuItems} navbar>
              <NavItem className={styles.menuItem}>
                {this.props.isAuth && (
                  <NvLink to="/recordings/newRecording">
                    <NavLink>New Recording</NavLink>
                  </NvLink>
                )}
              </NavItem>
              <NavItem className={styles.menuItem}>
                {this.props.isAuth && (
                  <NvLink to="/recordings/pastRecordings">
                    <NavLink>Past Recordings</NavLink>
                  </NvLink>
                )}
              </NavItem>
              {!this.props.isAuth ? unauthLogin : loginDropdown}
            </Nav>
          </Collapse>
        </Container>
      </Navbar>
    );
  }
}

const mapStateToProps = state => ({
  isAuth: state.user.isAuth
});

const mapDispatchToProps = dispatch => ({
  logout: () => dispatch(logout())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Header);
