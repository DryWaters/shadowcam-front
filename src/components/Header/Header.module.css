.headerContainer {
  box-shadow: #333 5px 10px;
}

.logo {
  font-family: "Righteous", cursive;
  font-size: 2em;
  color: white;
}

.logoLink {
  padding-bottom: 10px;
}

.logoLink:hover {
    color: white;
    text-decoration: none !important;
}

.menuItems {
  display: flex;
  flex: 1;
  justify-content: space-between;
  margin-left: 20%;
}

.menuItem {
  transition: border-bottom 0.25s ease-in-out;
  cursor: pointer;
}

.menuItem:after {
  display: block;
  content: "";
  border-bottom: solid 3px #8dc24c;
  transform: scaleX(0);
  transition: transform 250ms ease-in-out;
  transform-origin: 0% 100%;
}

.menuItem:hover:after {
  transform: scaleX(1);
  transform-origin: 0% 50%;
}

@media only screen and (max-width: 760px) {

  .menuItems {
    margin-left: 10%;
  }

  .menuItem:after {
    display: none;
  }
  .menuItem:hover:after {
    transform: none;
  }
}
