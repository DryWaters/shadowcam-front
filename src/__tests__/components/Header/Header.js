import React from "react";
import { shallow } from "enzyme";
import { Header } from "../../../components/Header/Header";
import { NavLink, Link } from "react-router-dom";
import { DropdownItem, Collapse, NavbarToggler } from "reactstrap";

describe("Header render", () => {
  it("Should render Header component without crashing", () => {
    shallow(<Header />);
  });

  it("Should match Header component snapshot", () => {
    const wrapper = shallow(<Header />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe("Header authorization tests", () => {
  it("Should not render links if not authorized", () => {
    const wrapper = shallow(<Header isAuth={false} />);
    expect(
      wrapper.contains(<NavLink to="/newRecording">New Recording</NavLink>).not
    );
    expect(
      wrapper.contains(<NavLink to="/pastRecordings">New Recording</NavLink>)
        .not
    );
    expect(
      wrapper.contains(<Link to="/account/profile">Account Settings</Link>).not
    );
    expect(wrapper.contains(<DropdownItem>Logout</DropdownItem>).not);
  });
  it("Should render links if authorized", () => {
    const wrapper = shallow(<Header isAuth />);
    expect(
      wrapper.contains(<NavLink to="/newRecording">New Recording</NavLink>)
    );
    expect(
      wrapper.contains(<NavLink to="/pastRecordings">New Recording</NavLink>)
    );
    expect(
      wrapper.contains(<Link to="/account/profile">Account Settings</Link>)
    );
    expect(wrapper.contains(<DropdownItem>Logout</DropdownItem>));
  });
});

describe("Header toggle dropdown working", () => {
  it("Should drop down toggle account menu when clicked", () => {
    const wrapper = shallow(<Header isAuth />);
    expect(wrapper.contains(<Collapse isOpen={false} navbar />));
    wrapper.find(NavbarToggler).simulate("click");
    expect(wrapper.contains(<Collapse isOpen={true} navbar />))
  });

  it("Should toggle menu when toggle method is executed", () => {
    const wrapper = shallow(<Header isAuth />);
    const instance = wrapper.instance();
    expect(wrapper.contains(<Collapse isOpen={false} navbar />));
    instance.toggle();
    expect(wrapper.contains(<Collapse isOpen={true} navbar />))
  });
});
