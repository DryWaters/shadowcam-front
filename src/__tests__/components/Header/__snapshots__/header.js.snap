// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header render Should match Header component snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Header />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Container
        tag="div"
      >
        <Link
          className="logoLink"
          replace={false}
          to="/"
        >
          <img
            alt="Logo"
            src="boxer-header.png"
          />
          <span
            className="logo"
          >
            ShadowCam
          </span>
        </Link>
        <NavbarToggler
          onClick={[Function]}
          tag="button"
          type="button"
        />
        <Collapse
          appear={false}
          enter={true}
          exit={true}
          in={false}
          isOpen={false}
          mountOnEnter={false}
          navbar={true}
          onEnter={[Function]}
          onEntered={[Function]}
          onEntering={[Function]}
          onExit={[Function]}
          onExited={[Function]}
          onExiting={[Function]}
          tag="div"
          timeout={350}
          unmountOnExit={false}
        >
          <Nav
            className="menuItems"
            navbar={true}
            tag="ul"
            vertical={false}
          >
            <NavItem
              className="menuItem"
              tag="li"
            />
            <NavItem
              className="menuItem"
              tag="li"
            />
            <Link
              replace={false}
              to="/account/login"
            >
              <Button
                color="secondary"
                tag="button"
              >
                Login
              </Button>
            </Link>
          </Nav>
        </Collapse>
      </Container>,
      "className": "headerContainer",
      "color": "dark",
      "dark": true,
      "expand": "md",
      "tag": "nav",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Link
            className="logoLink"
            replace={false}
            to="/"
          >
            <img
              alt="Logo"
              src="boxer-header.png"
            />
            <span
              className="logo"
            >
              ShadowCam
            </span>
          </Link>,
          <NavbarToggler
            onClick={[Function]}
            tag="button"
            type="button"
          />,
          <Collapse
            appear={false}
            enter={true}
            exit={true}
            in={false}
            isOpen={false}
            mountOnEnter={false}
            navbar={true}
            onEnter={[Function]}
            onEntered={[Function]}
            onEntering={[Function]}
            onExit={[Function]}
            onExited={[Function]}
            onExiting={[Function]}
            tag="div"
            timeout={350}
            unmountOnExit={false}
          >
            <Nav
              className="menuItems"
              navbar={true}
              tag="ul"
              vertical={false}
            >
              <NavItem
                className="menuItem"
                tag="li"
              />
              <NavItem
                className="menuItem"
                tag="li"
              />
              <Link
                replace={false}
                to="/account/login"
              >
                <Button
                  color="secondary"
                  tag="button"
                >
                  Login
                </Button>
              </Link>
            </Nav>
          </Collapse>,
        ],
        "tag": "div",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <img
                alt="Logo"
                src="boxer-header.png"
              />,
              <span
                className="logo"
              >
                ShadowCam
              </span>,
            ],
            "className": "logoLink",
            "replace": false,
            "to": "/",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "Logo",
                "src": "boxer-header.png",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "ShadowCam",
                "className": "logo",
              },
              "ref": null,
              "rendered": "ShadowCam",
              "type": "span",
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onClick": [Function],
            "tag": "button",
            "type": "button",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "appear": false,
            "children": <Nav
              className="menuItems"
              navbar={true}
              tag="ul"
              vertical={false}
            >
              <NavItem
                className="menuItem"
                tag="li"
              />
              <NavItem
                className="menuItem"
                tag="li"
              />
              <Link
                replace={false}
                to="/account/login"
              >
                <Button
                  color="secondary"
                  tag="button"
                >
                  Login
                </Button>
              </Link>
            </Nav>,
            "enter": true,
            "exit": true,
            "in": false,
            "isOpen": false,
            "mountOnEnter": false,
            "navbar": true,
            "onEnter": [Function],
            "onEntered": [Function],
            "onEntering": [Function],
            "onExit": [Function],
            "onExited": [Function],
            "onExiting": [Function],
            "tag": "div",
            "timeout": 350,
            "unmountOnExit": false,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <NavItem
                  className="menuItem"
                  tag="li"
                />,
                <NavItem
                  className="menuItem"
                  tag="li"
                />,
                <Link
                  replace={false}
                  to="/account/login"
                >
                  <Button
                    color="secondary"
                    tag="button"
                  >
                    Login
                  </Button>
                </Link>,
              ],
              "className": "menuItems",
              "navbar": true,
              "tag": "ul",
              "vertical": false,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": undefined,
                  "className": "menuItem",
                  "tag": "li",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": undefined,
                  "className": "menuItem",
                  "tag": "li",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <Button
                    color="secondary"
                    tag="button"
                  >
                    Login
                  </Button>,
                  "replace": false,
                  "to": "/account/login",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "Login",
                    "color": "secondary",
                    "tag": "button",
                  },
                  "ref": null,
                  "rendered": "Login",
                  "type": [Function],
                },
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Container
          tag="div"
        >
          <Link
            className="logoLink"
            replace={false}
            to="/"
          >
            <img
              alt="Logo"
              src="boxer-header.png"
            />
            <span
              className="logo"
            >
              ShadowCam
            </span>
          </Link>
          <NavbarToggler
            onClick={[Function]}
            tag="button"
            type="button"
          />
          <Collapse
            appear={false}
            enter={true}
            exit={true}
            in={false}
            isOpen={false}
            mountOnEnter={false}
            navbar={true}
            onEnter={[Function]}
            onEntered={[Function]}
            onEntering={[Function]}
            onExit={[Function]}
            onExited={[Function]}
            onExiting={[Function]}
            tag="div"
            timeout={350}
            unmountOnExit={false}
          >
            <Nav
              className="menuItems"
              navbar={true}
              tag="ul"
              vertical={false}
            >
              <NavItem
                className="menuItem"
                tag="li"
              />
              <NavItem
                className="menuItem"
                tag="li"
              />
              <Link
                replace={false}
                to="/account/login"
              >
                <Button
                  color="secondary"
                  tag="button"
                >
                  Login
                </Button>
              </Link>
            </Nav>
          </Collapse>
        </Container>,
        "className": "headerContainer",
        "color": "dark",
        "dark": true,
        "expand": "md",
        "tag": "nav",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Link
              className="logoLink"
              replace={false}
              to="/"
            >
              <img
                alt="Logo"
                src="boxer-header.png"
              />
              <span
                className="logo"
              >
                ShadowCam
              </span>
            </Link>,
            <NavbarToggler
              onClick={[Function]}
              tag="button"
              type="button"
            />,
            <Collapse
              appear={false}
              enter={true}
              exit={true}
              in={false}
              isOpen={false}
              mountOnEnter={false}
              navbar={true}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              tag="div"
              timeout={350}
              unmountOnExit={false}
            >
              <Nav
                className="menuItems"
                navbar={true}
                tag="ul"
                vertical={false}
              >
                <NavItem
                  className="menuItem"
                  tag="li"
                />
                <NavItem
                  className="menuItem"
                  tag="li"
                />
                <Link
                  replace={false}
                  to="/account/login"
                >
                  <Button
                    color="secondary"
                    tag="button"
                  >
                    Login
                  </Button>
                </Link>
              </Nav>
            </Collapse>,
          ],
          "tag": "div",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <img
                  alt="Logo"
                  src="boxer-header.png"
                />,
                <span
                  className="logo"
                >
                  ShadowCam
                </span>,
              ],
              "className": "logoLink",
              "replace": false,
              "to": "/",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "Logo",
                  "src": "boxer-header.png",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "ShadowCam",
                  "className": "logo",
                },
                "ref": null,
                "rendered": "ShadowCam",
                "type": "span",
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onClick": [Function],
              "tag": "button",
              "type": "button",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "appear": false,
              "children": <Nav
                className="menuItems"
                navbar={true}
                tag="ul"
                vertical={false}
              >
                <NavItem
                  className="menuItem"
                  tag="li"
                />
                <NavItem
                  className="menuItem"
                  tag="li"
                />
                <Link
                  replace={false}
                  to="/account/login"
                >
                  <Button
                    color="secondary"
                    tag="button"
                  >
                    Login
                  </Button>
                </Link>
              </Nav>,
              "enter": true,
              "exit": true,
              "in": false,
              "isOpen": false,
              "mountOnEnter": false,
              "navbar": true,
              "onEnter": [Function],
              "onEntered": [Function],
              "onEntering": [Function],
              "onExit": [Function],
              "onExited": [Function],
              "onExiting": [Function],
              "tag": "div",
              "timeout": 350,
              "unmountOnExit": false,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <NavItem
                    className="menuItem"
                    tag="li"
                  />,
                  <NavItem
                    className="menuItem"
                    tag="li"
                  />,
                  <Link
                    replace={false}
                    to="/account/login"
                  >
                    <Button
                      color="secondary"
                      tag="button"
                    >
                      Login
                    </Button>
                  </Link>,
                ],
                "className": "menuItems",
                "navbar": true,
                "tag": "ul",
                "vertical": false,
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": undefined,
                    "className": "menuItem",
                    "tag": "li",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": undefined,
                    "className": "menuItem",
                    "tag": "li",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": <Button
                      color="secondary"
                      tag="button"
                    >
                      Login
                    </Button>,
                    "replace": false,
                    "to": "/account/login",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": "Login",
                      "color": "secondary",
                      "tag": "button",
                    },
                    "ref": null,
                    "rendered": "Login",
                    "type": [Function],
                  },
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
